### ISSUE TITLE
Plug-in API for custom filters / effects

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#64]
**As a user, I want to apply custom filters to my images.**

For example to add glitch effects like this, pretty simple to implement with existing libraries: https://snorpey.github.io/jpg-glitch/

An option could be to run all filters in the browser only using JavaScript / Webassembly (uploading the result might take too long though for some users / internet connections). Let's think about it.

![download](https://user-images.githubusercontent.com/301686/48333591-51483f80-e658-11e8-881b-682dcc24530f.png)

**Comment by 0chroma** [comment]
I'd be interested in working on this for beauty filters!

**Comment by lastzero** [comment]
@0chroma That's great! I suggest you take a look at our existing backend APIs / frontend components and then suggest changes needed to implement this:
- https://pkg.go.dev/github.com/photoprism/photoprism/internal/api
- https://github.com/photoprism/photoprism/tree/develop/internal/api
- https://github.com/photoprism/photoprism/blob/develop/frontend/src/common/viewer.js

**Comment by 0chroma** [comment]
I thought a lil more about this, I think it might be better if I worked on something like this as a native app. It's common to take photos/video directly with a beauty filter app rather than postprocessing it, which seems outside of the scope of photoprism

that said, it would be neat if photoprism could detect that case or integrate with editor apps, then group it in with the original it came from. but thats pretty difficult to implement haha

**Comment by lastzero** [comment]
Stacking files actually isn't very complicated if you add / use a unique content identifier embedded into the metadata. This will work even if the files have different names and/or paths.

**Comment by 0chroma** [comment]
right, I'm more thinking closed source apps like snow which dont always preserve that metadata. It's frustrating.
